File: ChatServerCS/ChatHub.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Collections.Concurrent;
using Microsoft.AspNet.SignalR;

namespace ChatServerCS
{
    public class ChatHub : Hub<IClient>
    {
        private static ConcurrentDictionary<string, Utente> ChatClients = new ConcurrentDictionary<string, Utente>();

        public override Task OnDisconnected(bool stopCalled)
        {
            var userName = ChatClients.SingleOrDefault((c) => c.Value.ID == Context.ConnectionId).Key;
            if (userName != null)
            {
                Clients.Others.DisconnessionePartecipante(userName);
                Console.WriteLine($"<> {userName} ha chiuso il client");
            }
            return base.OnDisconnected(stopCalled);
        }

        public override Task OnReconnected()
        {
            var nomeUtente = ChatClients.SingleOrDefault((c) => c.Value.ID == Context.ConnectionId).Key;
            if (nomeUtente != null)
            {
                Clients.Others.RiconnessionePartecipante(nomeUtente);
                Console.WriteLine($"== {nomeUtente} riconnesso");
            }
            return base.OnReconnected();
        }

        public List<Utente> Accesso(string nome, byte[] foto)
        {
            if (!ChatClients.ContainsKey(nome))
            {
                Console.WriteLine($"++ {nome} si è connesso");
                List<Utente> utenti = new List<Utente>(ChatClients.Values);
                Utente nuovoUtente = new Utente { Nome = nome, ID = Context.ConnectionId, imgProfilo = foto };
                var aggiunto = ChatClients.TryAdd(nome, nuovoUtente);
                if (!aggiunto) return null;
                Clients.CallerState.UserName = nome;
                Clients.Others.AccessoPartecipante(nuovoUtente);
                return utenti;
            }
            return null;
        }

        public void Logout()
        {
            var nome = Clients.CallerState.UserName;
            if (!string.IsNullOrEmpty(nome))
            {
                Utente client = new Utente();
                ChatClients.TryRemove(nome, out client);
                Clients.Others.LogoutPartecipante(nome);
                Console.WriteLine($"-- {nome} si è disconnesso");
            }
        }

        public void MsgTestoBroadcast(string messaggio)
        {
            var nome = Clients.CallerState.UserName;
            if (!string.IsNullOrEmpty(nome) && !string.IsNullOrEmpty(messaggio))
            {
                Clients.Others.MsgTestoBroadcast(nome, messaggio);
            }
        }

        public void MsgImmagineBroadcast(byte[] img)
        {
            var nome = Clients.CallerState.UserName;
            if (img != null)
            {
                Clients.Others.MsgImmagineBroadcast(nome, img);
            }
        }

        public void MsgTestoUnicast(string destinatario, string messaggio)
        {
            var mittente = Clients.CallerState.UserName;
            if (!string.IsNullOrEmpty(mittente) && destinatario != mittente &&
                !string.IsNullOrEmpty(messaggio) && ChatClients.ContainsKey(destinatario))
            {
                Utente client = new Utente();
                ChatClients.TryGetValue(destinatario, out client);
                Clients.Client(client.ID).MsgTestoUnicast(mittente, messaggio);
            }
        }

        public void MsgImmagineUnicast(string destinatario, byte[] img)
        {
            var mittente = Clients.CallerState.UserName;
            if (!string.IsNullOrEmpty(mittente) && destinatario != mittente &&
                img != null && ChatClients.ContainsKey(destinatario))
            {
                Utente client = new Utente();
                ChatClients.TryGetValue(destinatario, out client);
                Clients.Client(client.ID).MsgImmagineUnicast(mittente, img);
            }
        }

        public void StaScrivendo(string destinatario)
        {
            if (string.IsNullOrEmpty(destinatario)) return;
            var mittente = Clients.CallerState.UserName;
            Utente client = new Utente();
            ChatClients.TryGetValue(destinatario, out client);
            Clients.Client(client.ID).PartecipanteStaScrivendo(mittente);
        }
    }
}

File: ChatServerVB/ChatHub.vb
Imports System.Collections.Concurrent
Imports Microsoft.AspNet.SignalR

Public Class ChatHub
    Inherits Hub(Of IClient)

    Private Shared ChatClients As New ConcurrentDictionary(Of String, User)

    Public Overrides Function OnDisconnected(stopCalled As Boolean) As Task
        Dim nomeUtente = ChatClients.SingleOrDefault(Function(c) c.Value.ID = Context.ConnectionId).Key
        If nomeUtente IsNot Nothing Then
            Clients.Others.DisconnessionePartecipante(nomeUtente)
            Console.WriteLine($"<> {nomeUtente} ha chiuso il client")
        End If
        Return MyBase.OnDisconnected(stopCalled)
    End Function

    Public Overrides Function OnReconnected() As Task
        Dim nomeUtente = ChatClients.SingleOrDefault(Function(c) c.Value.ID = Context.ConnectionId).Key
        If nomeUtente IsNot Nothing Then
            Clients.Others.RiconnessionePartecipante(nomeUtente)
            Console.WriteLine($"== {nomeUtente} riconnesso")
        End If
        Return MyBase.OnReconnected()
    End Function

    Public Function Accesso(ByVal nome As String, ByVal foto As Byte()) As List(Of User)
        If Not ChatClients.ContainsKey(nome) Then
            Console.WriteLine($"++ {nome} accesso effettuato")
            Dim users As New List(Of User)(ChatClients.Values)
            Dim nuovoUtente As New User With {.Nome = nome, .ID = Context.ConnectionId, .imgProfilo = foto}
            Dim added = ChatClients.TryAdd(nome, nuovoUtente)
            If Not added Then Return Nothing
            Clients.CallerState.UserName = nome
            Clients.Others.LoginPartecipante(nuovoUtente)
            Return users
        End If
        Return Nothing
    End Function

    Public Sub Logout()
        Dim nome = Clients.CallerState.UserName
        If Not String.IsNullOrEmpty(nome) Then
            Dim client As New User
            ChatClients.TryRemove(nome, client)
            Clients.Others.LogoutPartecipante(nome)
            Console.WriteLine($"-- {nome} si è disconnesso")
        End If
    End Sub

    Public Sub MsgTestoBroadcast(messaggio As String)
        Dim nome = Clients.CallerState.UserName
        If Not String.IsNullOrEmpty(nome) AndAlso Not String.IsNullOrEmpty(messaggio) Then
            Clients.Others.MsgTestoBroadcast(nome, messaggio)
        End If
    End Sub

    Public Sub MsgImmagineBroadcast(img As Byte())
        Dim nome = Clients.CallerState.UserName
        If img IsNot Nothing Then
            Clients.Others.MsgImmagineBroadcast(nome, img)
        End If
    End Sub

    Public Sub MsgTestoUnicast(destinatario As String, messaggio As String)
        Dim mittente = Clients.CallerState.UserName
        If Not String.IsNullOrEmpty(mittente) AndAlso destinatario <> mittente AndAlso
           Not String.IsNullOrEmpty(messaggio) AndAlso ChatClients.Keys.Contains(destinatario) Then
            Dim client As New User
            ChatClients.TryGetValue(destinatario, client)
            Clients.Client(client.ID).MsgTestoUnicast(mittente, messaggio)
        End If
    End Sub

    Public Sub MsgImmagineUnicast(destinatario As String, img As Byte())
        Dim mittente = Clients.CallerState.UserName
        If Not String.IsNullOrEmpty(mittente) AndAlso destinatario <> mittente AndAlso
           img IsNot Nothing AndAlso ChatClients.Keys.Contains(destinatario) Then
            Dim client As New User
            ChatClients.TryGetValue(destinatario, client)
            Clients.Client(client.ID).MsgImmagineUnicast(mittente, img)
        End If
    End Sub

    Public Sub StaScrivendo(destinatario As String)
        If String.IsNullOrEmpty(destinatario) Then Exit Sub
        Dim mittente = Clients.CallerState.UserName
        Dim client As New User
        ChatClients.TryGetValue(destinatario, client)
        Clients.Client(client.ID).PartecipanteStaScrivendo(mittente)
    End Sub
End Class